File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.cpp
// g++ -std=c++17 -Wall -Wextra -g3 /Users/xiang13air/Program/Leetcode/Test.cpp -o /Users/xiang13air/Program/Leetcode/output/Test

// # Given: an integer array nums and an integer k
// # return: the k most frequent elements within the array.
// # thoughts: hash map + count {key = num : value = count}
// # Complexity: aim for O(n) time and O(n) space (where n is the size of the input array).
// '''
// EX: nums = [1, 1, 1, 2, 2, 3] , k = 2
// Sol 1: 極致Pythonic!!!（Tools）
// Sol 2: 實作Sorting
// Sol 3: Min-Heap(前k大)— 🔥C++預設是max-heap!!!
// Sol 4: Bucket Sort
// '''

// # include <iostream>
// # include <vector>
// # include <unordered_map>
// # include <queue>
// # include <algorithm>
// using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // '''
        // Sol 2: 實作Sorting
        // '''
        // unordered_map<int, int> count; // 先計數
        // for (int& num : nums) {
        //     count[num] ++;
        // }
        // vector<pair<int, int>>arr; // num跟freq換位
        // for (const auto& p : count) {      
        //     arr.push_back({p.second, p.first});      
        // }
        // sort(arr.rbegin(), arr.rend()); // 再排序
        // vector<int> res;
        // for (int i = 0; i < k; i++) {
        //     res.push_back(arr[i].second); // ⚠️這時second是num，不是freq!
        // }
        // return res;
        
        // '''
        // Sol 3: Min-Heap(前k大)— 🔥C++預設是max-heap!!!
        // '''
        // unordered_map<int, int> count;
        // for (const auto& num : nums) {
        //     count[num]++;
        // }
        // priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> heap;
        // for (const auto& p : count) {
        //     heap.push({p.second, p.first}); // ⚠️heap.push()沒有_back()!
        //     while (heap.size() > k) {
        //         heap.pop();
        //     }
        // }
        // vector<int> res;
        // while (!heap.empty()) {
        //     res.push_back(heap.top().second);
        //     heap.pop();
        // }
        // return res;

        // Sol 4️⃣: Bucket Sort
        unordered_map<int, int> count;
        for (int num : nums) {
            count[num]++;
        }
        vector<vector<int>> bucket_freq(nums.size() + 1);
        for (const auto& p : count) {
            bucket_freq[p.second].push_back(p.first);
        }
        vector<int> res;
        for (int i = bucket_freq.size() -1; i > 0; --i) {
            for (int num : bucket_freq[i]) {
                res.push_back(num);
                if (res.size() == size_t(k)) {
                    return res;
                }
            }
        }
        return res;
    }
};

// int main()
// {
//     vector<int> nums = {1, 1, 1, 2, 2, 3};
//     int k = 2;

//     Solution sol;
//     vector<int> result = sol.topKFrequent(nums, k);

//     cout << "[";
//     for (int num : result) {
//         cout << num;
//         if (num != result.back()) {
//             cout << ", ";
//         }
//     }
//     cout << "]" << endl;


//     return 0;
// };

File: Data Structures & Algorithms/counting-bits/submission-1.py
# # # # # class Solution:
# # # # #     def countBits(self, n: int) -> List[int]:
# # # # #         res = []
# # # # #         for num in range(n + 1):
# # # # #             count = 0
# # # # #             for i in range(32):
# # # # #                 if num & (1 << i):
# # # # #                     count += 1
# # # # #             res.append(count)
# # # # #         return res

# # # # class Solution:
# # # #     def countBits(self, n: int) -> List[int]:
# # # #         res = []
# # # #         for num in range(n + 1):
# # # #             count = 0
# # # #             while num:
# # # #                 num &= (num -1)
# # # #                 count += 1
# # # #             res.append(count)
# # # #         return res


# # # class Solution:
# # #     def countBits(self, n:int) -> List[int]:
# # #         res = [0] * (n + 1)
# # #         for i in range(1, n + 1):
# # #             num = i
# # #             while num != 0:
# # #                 num &= (num - 1)
# # #                 res[i] += 1
# # #         return res

# # class Solution:
# #     def countBits(self, n:int) -> List[int]:
# #         res = [0] * (n + 1)
# #         for i in range(n + 1):
# #             res[i] = bin(i).count('1')
# #         return res

# class Solution:
#     def countBits(self, n:int) -> List[int]:
#         dp = [0] * (n + 1)
#         offset = 1
#         for i in range(1, n + 1):
#             if offset * 2 == i:
#                 offset = i
#             dp[i] = 1 + dp[i - offset]
#         return dp

class Solution:
    def countBits(self, n:int) -> List[int]:
        dp = [0] * (n + 1)
        for i in range(1, n + 1):
            dp[i] = dp[i >> 1] + (i & 1)
        return dp